### DISCUSSION TITLE
Reuse common code with mbtileserver and others

### DISCUSSION THREAD

**Discussion by nyurik** [maplibre/maplibre-rs#106] [Ideas]
Both `maplibre-rs` and `mbtileserver` use near-identical TileJSON struct and other code. I believe we should refactor the code such that they can be easily shared across multiple projects. Preliminary list of shareable features:
* `TileJSON` - use [georust/tilejson](https://github.com/georust/tilejson) lib for the serializable struct representing [tilejson spec](https://github.com/mapbox/tilejson-spec).
* `z/x/y` representation - integer representation of the tile coordinate, with validation, and max value per zoom. Also declares MAX_ZOOM. Could also include packed representation, e.g. single u32 value representing z 0..15.
* style language parsing

All of the above are fairly tightly coupled, so might make sense to have them as a single crate.

TBD:
* crate name
* code location -- `maplibre-rs` repo (my preference) or a separate repo?
* research existing crates - there might be already something very close that we can use instead?

**Comment by maxammann** [comment]
Right extracting functionality is not a major priority for maplibre-rs. Right now its very handy to have everything at one place while things are moving very quickly. It could make sense to extract tilejson, but at the same time it could make things more complicated without a real benefit.

For example it would not be possible to implement an arbitrary external `trait` for the tilejson struct if it does not live in the `maplibre-rs` core crate.

So I think we should discuss this again in about half a year, when things have settled a little bit more.

**Reply by nyurik** [reply]
@maxammann thx, it does make sense to keep it close to `maplibre-rs` as it is clearly the primary project for all such things. Do you think it would be too much hassle to refactor it out into a separate crate though? (keeping in the same repo)  Obviously we could start with copy/pasting with the intention of merging later I guess, i.e. keep it in a separate module like you did with tilejson code.  Speaking of which, there are [some issues](https://github.com/georust/tilejson/issues) with that code that i will try to address.

**Reply by maxammann** [reply]
> @maxammann thx, it does make sense to keep it close to `maplibre-rs` as it is clearly the primary project for all such things. Do you think it would be too much hassle to refactor it out into a separate crate though? (keeping in the same repo)  Obviously we could start with copy/pasting with the intention of merging later I guess, i.e. keep it in a separate module like you did with tilejson code.  Speaking of which, there are [some issues](https://github.com/georust/tilejson/issues) with that code that i will try to address.

I think it should work, but the undocumented guideline currently is that only crates which start with maplibre- should be published.
 Nonetheless its probably better to do this in about 3-6 months when we changes to the core structs are less likely.
